app/src/main/java/mod/hilal/saif/activities/tools/AddCustomComponentActivity.java

```diff
@@ -0,0 +1,258 @@
+            || componentBuildClass.getText().toString().isEmpty();
+    }
+
+    private void save() {
+        ArrayList<HashMap<String, Object>> list = new ArrayList<>();
+        if (FileUtil.isExistFile(path)) {
+            list = new Gson().fromJson(FileUtil.readFile(path), Helper.TYPE_MAP_LIST);
+        }
+        HashMap<String, Object> map = new HashMap<>();
+        if (isEditMode) {
+            map = list.get(position);
+        }
+        map.put("name", componentName.getText().toString());
+        map.put("id", componentId.getText().toString());
+        map.put("icon", componentIcon.getText().toString());
+        map.put("varName", componentVarTypeName.getText().toString());
+        map.put("typeName", componentTypeName.getText().toString());
+        map.put("buildClass", componentBuildClass.getText().toString());
+        map.put("class", componentTypeClass.getText().toString());
+        map.put("description", componentDesc.getText().toString());
+        map.put("url", componentDocUrl.getText().toString());
+        map.put("additionalVar", componentAddVar.getText().toString());
+        map.put("defineAdditionalVar", componentDefineAddVar.getText().toString());
+        map.put("imports", componentImports.getText().toString());
+        if (!isEditMode) {
+            list.add(map);
+        }
+        FileUtil.writeFile(path, new Gson().toJson(list));
+        SketchwareUtil.toast(Helper.getResString(R.string.common_word_saved));
+        finish();
+    }
+
+    private void showIconSelectorDialog() {
+        new IconSelectorDialog(this, componentIcon).show();
+    }
+
+    private void showFilePickerDialog() {
+        SketchFilePickerDialog dialog = new SketchFilePickerDialog(this)
+            .addAllowedExtension("json")
+            .setFilePath(FileUtil.getExternalStorageDir())
+            .addOnFileSelectedListener((SketchFilePickerDialog _dialog, File file) -> {
+                try {
+                    selectComponentToImport(file.getAbsolutePath());
+                } catch (Exception e) {
+                    SketchwareUtil.toastError(Helper.getResString(R.string.publish_message_dialog_invalid_json));
+                }
+                _dialog.dismiss();
+            })
+            .addOnDismissListener((SketchFilePickerDialog _dialog) -> _dialog.dismiss());
+        dialog.setTitle(Helper.getResString(R.string.common_word_import));
+        dialog.setIcon(R.drawable.file_48_blue);
+        dialog.show();
+    }
+
+    private void selectComponentToImport(String path) {
+        String text = FileUtil.readFile(path);
+        if (text.isEmpty() || text.equals("[]")) {
+            SketchwareUtil.toastError(Helper.getResString(R.string.common_message_selected_file_empty));
+            return;
+        }
+
+        ArrayList<HashMap<String, Object>> list = new Gson().fromJson(text, Helper.TYPE_MAP_LIST);
+        if (list.isEmpty() || !ComponentsHandler.isValidComponentList(list)) {
```

Comment:
this new method is already used here, but you really create it later. please fix the order, e.g. by just creating a new commit before this saying something like _add new component validation methods_. otherwise this commit won't compile on its own which isn't a good thing.